Rebus.AmazonS3/AmazonS3/AmazonS3SubscriptionsStorage.cs

```diff
@@ -0,0 +1,219 @@
+using Amazon.S3;
+using Rebus.Logging;
+using System.Collections.Generic;
+using System.Linq;
+using Amazon.S3.Model;
+using System;
+using System.Net;
+using Rebus.Exceptions;
+using Amazon.S3.Util;
+using Rebus.Config;
+
+namespace Rebus.AmazonS3.AmazonS3
+{
+    /// <summary>
+    /// Implementation of <see cref="ISubscriptionStorage"/> that stores subscriptions in Amazon S3
+    /// </summary>
+    internal class AmazonS3SubscriptionsStorage : ISubscriptionStorage
+    {
+        private readonly AWSCredentials _awsCredentials;
+        private readonly AmazonS3Config _awsConfig;
+        private readonly AmazonS3DataBusOptions _options;
+        private readonly Func<IAmazonS3> _amazonS3Factory;
+        private readonly ILog _log;
+
+        public AmazonS3SubscriptionsStorage(AWSCredentials credentials, AmazonS3Config amazonS3Config, AmazonS3DataBusOptions options, IRebusLoggerFactory rebusLoggerFactory)
+            : this(rebusLoggerFactory)
+        {
+            _awsCredentials = credentials ?? throw new ArgumentNullException(nameof(credentials));
+            _awsConfig = amazonS3Config ?? throw new ArgumentNullException(nameof(amazonS3Config));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        public AmazonS3SubscriptionsStorage(Func<IAmazonS3> amazonS3Factory, AmazonS3DataBusOptions options, IRebusLoggerFactory rebusLoggerFactory)
+            : this(rebusLoggerFactory)
+        {
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _amazonS3Factory = amazonS3Factory;
+        }
+
+        private AmazonS3SubscriptionsStorage(IRebusLoggerFactory rebusLoggerFactory)
+        {
+            _log = rebusLoggerFactory?.GetLogger<AmazonS3SubscriptionsStorage>() ?? throw new ArgumentNullException(nameof(rebusLoggerFactory));
+        }
+
+        public bool IsCentralized
+        {
+            get { return true; }
+        }
+
+        public async Task<string[]> GetSubscriberAddresses(string topic)
+        {
+            await EnsureBucketExistAsync();
+
+            var subscriptions = await GetSubscriptions(topic);
+
+            return subscriptions.ToArray();
+        }
+
+        public async Task RegisterSubscriber(string topic, string subscriberAddress)
+        {
+            await EnsureBucketExistAsync();
+
+            string key = $"{topic}/{subscriberAddress}";
+
+            await PutObjectAsync(key);
+        }
+
+        public async Task UnregisterSubscriber(string topic, string subscriberAddress)
+        {
+            await EnsureBucketExistAsync();
+
+            string key = $"{topic}/{subscriberAddress}";
+
+            await DeleteObjectAsync(key);
+        }
+
+        private async Task<IList<string>> GetSubscriptions(string topic)
+        {
+            var topicKeys = await GetKeysForPrefix(topic);
+
+            var keys = topicKeys.Select(k => k.Substring(k.LastIndexOf('/') + 1));
```

Comment:
What happens if the `subscriberAddress` contains a `/` character?

I assume this could be likely to happen, say if the subscriber address is a SQS queue URL or similar? (_Haven't looked into exactly how Rebus stores SQS subscriptions, but I would assume it stores the queue URL_)

Might be possible to mitigate by assuming the form `{topic}/{subscriberAddress}` and then chopping of the known topic from the front rather than looking for the last `/` separator?

Maybe even just `k => k.Substring(topic.Length + 1)`?

Reply:
Might also be good to have a unit test specific for this project, that tests the above scenario both when `subscriberAddress` and `topic` can contain the special character `/`?